qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -218,88 +196,115 @@ def add_data(
                 LOG.warning(
                     "Adding a job from a backend (%s) that is different "
                     "than the current backend (%s). "
                     "The new backend will be used, but "
                     "service is not changed if one already exists.",
-                    data.backend(),
+                    job.backend(),
                     self.backend,
                 )
-            self._backend = data.backend()
+            self._backend = job.backend()
             if not self._service:
                 self._set_service_from_backend(self._backend)
 
-            self._jobs[data.job_id()] = data
-            job_kwargs = {
-                "job": data,
-                "job_done_callback": post_processing_callback,
-                "timeout": timeout,
-                **kwargs,
-            }
-            self._job_futures.append(
-                (
-                    job_kwargs,
-                    self._executor.submit(self._wait_for_job, **job_kwargs),
-                )
+            self._jobs[job.job_id()] = job
+
+        job_kwargs = {
+            "jobs": job_data,
+            "timeout": timeout,
+        }
+        self._job_futures.append(
+            (
+                job_kwargs,
+                self._executor.submit(self._add_jobs_result, **job_kwargs),
             )
-            if self.auto_save:
-                self.save_metadata()
-            return
+        )
+        if self.auto_save:
+            self.save_metadata()
 
-        if isinstance(data, dict):
-            self._add_single_data(data)
-        elif isinstance(data, Result):
-            self._add_result_data(data)
-        elif isinstance(data, list):
-            for dat in data:
-                self.add_data(dat)
+    def add_processing_callback(
+        self,
+        callback: Callable,
+        **kwargs: Any,
+    ):
+        """Add processing callback for after experiment job has run.
+
+        This method waits for the last set of jobs to finish and calls
+        the `callback` function asynchronously.
+
+        Note:
+            This method is not thread safe and should not be called by the
+            ` callback` function.
+
+        Args:
+            callback: Callback function invoked when job finishes successfully.
+                The following positional arguments are provided to the callback function:
+                    * This ``DbExperimentData`` object.
+                    * Additional keyword arguments passed to this method.
+            **kwargs: Keyword arguments to be passed to the callback function.
+        """
+        # Check if there are no futures to wait on
+        if not self._job_futures:
+            callback(self, **kwargs)
         else:
-            raise TypeError(f"Invalid data type {type(data)}.")
+            # Get the last added future and add a done callback
+            _, future = self._job_futures[-1]
```

Comment:
What happens in `self._job_futures[:-1]`? Submission of jobs and their completion are always in the same order? I guess this is not guaranteed because job is scheduled according to some formula and it prioritizes jobs with cost, i.e. number of circuits, shots, etc...